index.d.ts

```diff
@@ -0,0 +1,299 @@
+  export function bimap<LA, LB, RA, RB>(lmapper: (reason: LA) => LB): (rmapper: (value: RA) => RB) => (source: Future<LA, RA>) => Future<LB, RB>
+
+  /* Wait for both Futures to resolve in parallel. See https://github.com/fluture-js/Fluture#both */
+  export function both<L, A, B>(left: Future<L, A>, right: Future<L, B>): Future<L, [A, B]>
+  export function both<L, A, B>(left: Future<L, A>): (right: Future<L, B>) => Future<L, [A, B]>
+
+  /* Cache the outcome of the given Future. See https://github.com/fluture-js/Fluture#cache */
+  export function cache<L, R>(source: Future<L, R>): Future<L, R>
+
+  /* Create a Chain using the resolution value of the given Chain. See https://github.com/fluture-js/Fluture#chain */
+  export function chain<L, RA, RB>(mapper: (value: RA) => Future<L, RB>, source: Future<L, RA>): Future<L, RB>
+  export function chain<L, RA, RB>(mapper: (value: RA) => Future<L, RB>): (source: Future<L, RA>) => Future<L, RB>
+
+  /* Fork the given Future into a Node-style callback. See https://github.com/fluture-js/Fluture#done */
+  export function done<L, R>(callback: Nodeback<L, R>, source: Future<L, R>): Cancel
+  export function done<L, R>(callback: Nodeback<L, R>): (source: Future<L, R>) => Cancel
+
+  /* Encase the given function such that it returns a Future of its return value. See https://github.com/fluture-js/Fluture#encase */
+  export function encase<L, R, A>(fn: (a: A) => R, a: A): Future<L, R>
+  export function encase<L, R, A>(fn: (a: A) => R): {(a: A): Future<L, R>}
+
+  /* Encase the given function such that it returns a Future of its return value. See https://github.com/fluture-js/Fluture#encase */
+  export function encase2<L, R, A, B>(fn: (a: A, b: B) => R, a: A, b: B): Future<L, R>
+  export function encase2<L, R, A, B>(fn: (a: A, b: B) => R, a: A): (b: B) => Future<L, R>
+  export function encase2<L, R, A, B>(fn: (a: A, b: B) => R): (a: A, b: B) => Future<L, R>
+  export function encase2<L, R, A, B>(fn: (a: A, b: B) => R): (a: A) => (b: B) => Future<L, R>
+
+  /* Encase the given function such that it returns a Future of its return value. See https://github.com/fluture-js/Fluture#encase */
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R, a: A, b: B, c: C): Future<L, R>
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R, a: A, b: B): (c: C) => Future<L, R>
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R, a: A): (b: B, c: C) => Future<L, R>
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R, a: A): (b: B) => (c: C) => Future<L, R>
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R): (a: A, b: B, c: C) => Future<L, R>
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R): (a: A) => (b: B, c: C) => Future<L, R>
+  export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R): (a: A) => (b: B) => (c: C) => Future<L, R>
```

Comment:
Actually, when annotating `a:string, b:string, c:string`, TypeScript will let it pass.

Unfortunately there's another, bigger, problem. The return value of these functions is not overloaded. TypeScript considers the following two definitions a "conflict" because they both have the exact same types of arguments, but "differing" return values:

```ts
export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R): (a: A, b: B, c: C) => Future<L, R>
export function encase3<L, R, A, B, C>(fn: (a: A, b: B, c: C) => R): (a: A) => (b: B, c: C) => Future<L, R>
```

And it deals with it by simply ignoring the second.